py/autoevals/oai.py

```diff
@@ -26,45 +29,91 @@ def open_cache():
     return _CONN
 
 
+def log_openai_request(span, input_args, response, **kwargs):
+    span = span or current_span()
+    if not span:
+        return
+
+    input = input_args.pop("messages")
+    span.log(
+        metrics={
+            "tokens": response["usage"]["total_tokens"],
+            "prompt_tokens": response["usage"]["prompt_tokens"],
+            "completion_tokens": response["usage"]["completion_tokens"],
+        },
+        metadata={**input_args, **kwargs},
+        input=input,
+        output=response["choices"][0],
+    )
+
+
+@traced(name="OpenAI Completion")
 def run_cached_request(Completion=None, **kwargs):
-    if Completion is None:
-        # OpenAI is very slow to import, so we only do it if we need it
-        import openai
+    # OpenAI is very slow to import, so we only do it if we need it
+    import openai
 
+    if Completion is None:
         Completion = openai.Completion
 
     param_key = json.dumps(kwargs)
     conn = open_cache()
     cursor = conn.cursor()
     resp = cursor.execute("""SELECT response FROM "cache" WHERE params=?""", [param_key]).fetchone()
+    cached = False
+    retries = 0
     if resp:
-        return json.loads(resp[0])
-
-    resp = Completion.create(**kwargs).to_dict()
-
-    cursor.execute("""INSERT INTO "cache" VALUES (?, ?)""", [param_key, json.dumps(resp)])
-    conn.commit()
+        cached = True
+        resp = json.loads(resp[0])
+    else:
+        sleep_time = 0.1
+        while retries < 20:
+            try:
+                resp = Completion.create(**kwargs).to_dict()
+                break
+            except openai.error.RateLimitError:
+                sleep_time *= 1.5
+                time.sleep(sleep_time)
+                retries += 1
+
+        cursor.execute("""INSERT INTO "cache" VALUES (?, ?)""", [param_key, json.dumps(resp)])
+        conn.commit()
+
+    log_openai_request(current_span(), kwargs, resp, cached=cached)
 
     return resp
 
 
+@traced(name="OpenAI Completion")
 async def arun_cached_request(Completion=None, **kwargs):
-    if Completion is None:
-        # OpenAI is very slow to import, so we only do it if we need it
-        import openai
+    # OpenAI is very slow to import, so we only do it if we need it
+    import openai
 
+    if Completion is None:
         Completion = openai.Completion
 
     param_key = json.dumps(kwargs)
     conn = open_cache()
     cursor = conn.cursor()
     resp = cursor.execute("""SELECT response FROM "cache" WHERE params=?""", [param_key]).fetchone()
+    cached = False
```

Comment:
can we avoid code duplication with above?

Reply:
not easily, since we use a different sleep method, etc. here

let's reconsider later.